File: libs/moqtail-ts/src/model/control/subscribe_ok.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { ControlMessageType } from './constant'
import { LengthExceedsMaxError } from '../error/error'
import { MessageParameter, MessageParameters } from '../parameter/message_parameter'
import { TrackExtension } from '../extension_header/track_extension'

import { Location } from '../common/location'
import { GroupOrder } from './constant'
import { DeliveryTimeoutExtension } from '../extension_header/track_extension'
import { Expires } from '../parameter/message/expires'
import { LargestObject } from '../parameter/message/largest_object'
import { GroupOrderParam } from '../parameter/message/group_order_param'
import { DeliveryTimeout } from '../parameter/message/delivery_timeout'

export class SubscribeOk {
  requestId: bigint
  trackAlias: bigint
  parameters: MessageParameter[]
  trackExtensions: TrackExtension[]

  private constructor(
    requestId: bigint,
    trackAlias: bigint,
    parameters: MessageParameter[],
    trackExtensions: TrackExtension[],
  ) {
    this.requestId = requestId
    this.trackAlias = trackAlias
    this.parameters = parameters
    this.trackExtensions = trackExtensions
  }

  static create(
    requestId: bigint,
    trackAlias: bigint,
    parameters: MessageParameter[],
    trackExtensions: TrackExtension[] = [],
  ): SubscribeOk {
    return new SubscribeOk(requestId, trackAlias, parameters, trackExtensions)
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.SubscribeOk)

    const payload = new ByteBuffer()
    payload.putVI(this.requestId)
    payload.putVI(this.trackAlias)
    payload.putVI(this.parameters.length)
    for (const param of this.parameters) {
      payload.putKeyValuePair(param.toKeyValuePair())
    }
    TrackExtension.serializeInto(this.trackExtensions, payload)
    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('SubscribeOk::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): SubscribeOk {
    const requestId = buf.getVI()
    const trackAlias = buf.getVI()
    const paramCount = buf.getNumberVI()
    const rawParams = new Array(paramCount)
    for (let i = 0; i < paramCount; i++) {
      rawParams[i] = buf.getKeyValuePair()
    }
    const parameters = MessageParameters.fromKeyValuePairs(rawParams)
    const trackExtensions = TrackExtension.deserializeAll(buf)
    return new SubscribeOk(requestId, trackAlias, parameters, trackExtensions)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('SubscribeOk', () => {
    test('roundtrip', () => {
      const requestId = 145136n
      const trackAlias = 999n
      const largestLocation = new Location(34n, 0n)
      const parameters = [
        new Expires(16n),
        new GroupOrderParam(GroupOrder.Ascending),
        new LargestObject(largestLocation),
        new DeliveryTimeout(100n),
      ]
      const subscribeOk = SubscribeOk.create(requestId, trackAlias, parameters)
      const frozen = subscribeOk.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.SubscribeOk))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = SubscribeOk.parsePayload(frozen)
      expect(deserialized.requestId).toBe(subscribeOk.requestId)
      expect(deserialized.trackAlias).toBe(subscribeOk.trackAlias)
      expect(deserialized.parameters.length).toBe(4)
      expect(deserialized.trackExtensions.length).toBe(0)
      expect(frozen.remaining).toBe(0)
    })

    test('roundtrip with track extensions', () => {
      const subscribeOk = SubscribeOk.create(
        145136n,
        999n,
        [new Expires(16n), new LargestObject(new Location(34n, 0n))],
        [new DeliveryTimeoutExtension(5000n)],
      )
      const frozen = subscribeOk.serialize()
      frozen.getVI() // message type
      const msgLength = frozen.getU16()
      const payload = new FrozenByteBuffer(frozen.getBytes(msgLength))
      const deserialized = SubscribeOk.parsePayload(payload)
      expect(deserialized.trackExtensions.length).toBe(1)
      expect(deserialized.trackExtensions[0]).toBeInstanceOf(DeliveryTimeoutExtension)
      expect(payload.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const requestId = 145136n
      const trackAlias = 999n
      const subscribeOk = SubscribeOk.create(requestId, trackAlias, [
        new Expires(16n),
        new LargestObject(new Location(34n, 0n)),
        new DeliveryTimeout(100n),
      ])
      const serialized = subscribeOk.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.SubscribeOk))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const payload = new FrozenByteBuffer(frozen.getBytes(msgLength))
      const deserialized = SubscribeOk.parsePayload(payload)
      expect(deserialized.requestId).toBe(subscribeOk.requestId)
      expect(deserialized.trackAlias).toBe(subscribeOk.trackAlias)
      expect(deserialized.parameters.length).toBe(3)
      expect(payload.remaining).toBe(0)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const subscribeOk = SubscribeOk.create(145136n, 999n, [
        new Expires(16n),
        new LargestObject(new Location(34n, 0n)),
      ])
      const serialized = subscribeOk.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        SubscribeOk.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/subscribe_ok.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::constant::ControlMessageType;
use super::control_message::ControlMessageTrait;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use crate::model::parameter::message_parameter::{
  MessageParameter, deserialize_message_parameters, serialize_message_parameters,
};
use crate::model::property::track_property::{
  TrackProperty, deserialize_track_properties, serialize_track_properties,
};
use bytes::{BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct SubscribeOk {
  pub track_alias: u64,
  pub subscribe_parameters: Vec<MessageParameter>,
  pub track_properties: Vec<TrackProperty>,
}

impl SubscribeOk {
  pub fn new(
    track_alias: u64,
    subscribe_parameters: Vec<MessageParameter>,
    track_properties: Vec<TrackProperty>,
  ) -> Self {
    Self {
      track_alias,
      subscribe_parameters,
      track_properties,
    }
  }
}

impl ControlMessageTrait for SubscribeOk {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::SubscribeOk)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.track_alias)?;

    payload.put_vi(self.subscribe_parameters.len() as u64)?;
    payload.extend_from_slice(&serialize_message_parameters(&self.subscribe_parameters)?);

    // Track Properties (no length prefix; bounded by outer message Length field)
    payload.extend_from_slice(&serialize_track_properties(&self.track_properties)?);

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "SubscribeOk::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let track_alias = payload.get_vi()?;

    let param_count = payload.get_vi()?;
    let subscribe_parameters =
      deserialize_message_parameters(payload, param_count, ControlMessageType::SubscribeOk)?;

    // Track Properties: consume whatever remains in the payload
    let track_properties = deserialize_track_properties(payload)?;

    Ok(Box::new(SubscribeOk {
      track_alias,
      subscribe_parameters,
      track_properties,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::SubscribeOk
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::model::common::location::Location;
  use crate::model::control::constant::GroupOrder;
  use crate::model::property::track_property::TrackProperty;
  use bytes::Buf;

  #[test]
  fn test_roundtrip() {
    let mut subscribe_ok = SubscribeOk {
      track_alias: 0,
      subscribe_parameters: vec![
        MessageParameter::new_expires(16),
        MessageParameter::new_group_order(GroupOrder::Ascending),
        MessageParameter::new_largest_object(Location {
          group: 34,
          object: 0,
        }),
        MessageParameter::new_expires(100),
      ],
      track_properties: vec![],
    };
    // Wire encoding canonicalizes parameter order ascending by type (delta-encoding requirement).
    subscribe_ok
      .subscribe_parameters
      .sort_by_key(|p| p.type_value());

    let mut buf = subscribe_ok.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::SubscribeOk as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = SubscribeOk::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe_ok);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_roundtrip_with_track_properties() {
    let subscribe_ok = SubscribeOk {
      track_alias: 42,
      subscribe_parameters: vec![
        MessageParameter::new_expires(0),
        MessageParameter::new_group_order(GroupOrder::Descending),
      ],
      track_properties: vec![
        TrackProperty::ObjectDeliveryTimeout { timeout_ms: 500 },
        TrackProperty::DynamicGroups { enabled: true },
      ],
    };

    let mut buf = subscribe_ok.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::SubscribeOk as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = SubscribeOk::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, subscribe_ok);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_excess_roundtrip() {
    let subscribe_ok = SubscribeOk {
      track_alias: 89123u64,
      subscribe_parameters: vec![
        MessageParameter::new_expires(16),
        MessageParameter::new_largest_object(Location {
          group: 34,
          object: 0,
        }),
      ],
      track_properties: vec![],
    };

    let serialized = subscribe_ok.serialize().unwrap();
    let mut excess = BytesMut::new();
    excess.extend_from_slice(&serialized);
    excess.extend_from_slice(&[9u8, 1u8, 1u8]);
    let mut buf = excess.freeze();

    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::SubscribeOk as u64);
    let msg_length = buf.get_u16();

    assert_eq!(msg_length as usize, buf.remaining() - 3);
    let mut payload = buf.copy_to_bytes(msg_length as usize);
    let deserialized = SubscribeOk::parse_payload(&mut payload).unwrap();
    assert_eq!(*deserialized, subscribe_ok);
    assert!(!payload.has_remaining());
    assert_eq!(buf.chunk(), &[9u8, 1u8, 1u8]);
  }

  #[test]
  fn test_partial_message() {
    let subscribe_ok = SubscribeOk {
      track_alias: 1223u64,
      subscribe_parameters: vec![
        MessageParameter::new_expires(16),
        MessageParameter::new_largest_object(Location {
          group: 34,
          object: 0,
        }),
      ],
      track_properties: vec![],
    };
    let mut buf = subscribe_ok.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::SubscribeOk as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());

    let upper = buf.remaining() / 2;
    let mut partial = buf.slice(..upper);
    let deserialized = SubscribeOk::parse_payload(&mut partial);
    assert!(deserialized.is_err());
  }
}
